### PR TITLE
feat(indicators): Fibonacci Retracement overlay (Feature 25)

### PR DESCRIPTION
**Pull request by saikrishna-dev**
## What

Adds **Feature 25 — Fibonacci Retracement overlay**: compute the 7 standard retracement levels (0 / 23.6 / 38.2 / 50 / 61.8 / 78.6 / 100 %) from the most recent swing high/low in the chart window, return them in the `/predict` `indicators` payload, and draw them as optional dashed horizontal lines on the price chart behind a **FIB** toggle.

## Why

Fibonacci retracement was the last open item in the roadmap's *Signals & Indicators* backlog. It rounds out the technical-signals suite (ATR, Stochastic, ADX, OBV, divergences) with a widely-used support/resistance overlay, at near-zero cost — it rides inside the existing `/predict` response, no new endpoint or proxy.

## Changes

**Backend**
- `models.py` — new self-contained `calculate_fibonacci_levels(highs, lows, lookback=60)` helper: finds the swing high/low over the window, derives `direction` (`up`/`down` by swing order), and returns the 7 levels interpolated in ascending price order. Returns `None` on insufficient (<10 bars) or flat data, so a bad indicator never aborts `/predict` (same convention as the other helpers).
- `routers/predict.py` — computes fib in the Step 4b# advanced-signals block from the existing `highs`/`lows` lists and adds `indicators.fibonacci` (`null` when absent).

**Frontend**
- `types/index.ts` — `FibonacciLevels` interface + `fibonacci?` field on the indicators type.
- `AdvancedChart.tsx` — new `fibLevels` / `showFib` / `fibColor` props draw dashed amber `createPriceLine` rows labelled `"<pct>% <price>"`, mirroring the existing support/resistance overlay.
- `PriceChartCard.tsx` — **FIB** `ToggleButton` (rendered only on the 2Y view when fib data is present), default off, reset on symbol change; wired to `AdvancedChart` with `theme.palette.warning.main`.

**Tests / Roadmap**
- `backend/tests/test_fibonacci.py` — 5 pure unit tests (level ordering/bounds, direction up/down, insufficient-data → `None`, flat-series → `None`).
- Roadmap: Feature 25 moved to Shipped; backlog line removed.

## Reviewer notes

- The price chart is **lightweight-charts** (`AdvancedChart`), not Recharts — VWAP and support/resistance are drawn via `createPriceLine`, so the fib overlay follows that same prop-driven pattern rather than a Recharts `<ReferenceLine>`.
- Verified locally against a fresh dev server: `/predict` for AAPL returns 7 correct descending levels (swing_high 317.40 → swing_low 245.28, direction `up`); the FIB toggle is hidden by default, appears only with fib data, and toggles the overlay on/off cleanly with no console errors.
- Full backend suite green (**303 passed, 1 skipped**), `ruff check backend/` clean, frontend `lint` + `build` pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added Fibonacci Retracement overlay with automatic swing-high/low detection generating 7 standard retracement levels.
  * Toggleable dashed-line overlay on price charts to visualize retracement levels (available in 2Y interval view).

* **Tests**
  * Added comprehensive test coverage for Fibonacci level calculations, validation, and edge cases.

### CODE REVIEW

File: backend/models.py [hunk 1]
```diff
@@ -332,6 +332,42 @@ def calculate_atr(
...
+            return None
+        h = highs[-lookback:]
+        lo = lows[-lookback:]
+        swing_high = max(h)
+        swing_low = min(lo)
+        if swing_high <= swing_low:
+            return None
+        hi_idx = h.index(swing_high)
+        lo_idx = lo.index(swing_low)
+        direction = "up" if lo_idx < hi_idx else "down"
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/models.py]: Applied — added guards for mismatched `highs`/`lows` lengths and `lookback <= 0` (returns None instead of a possibly-wrong direction), and clamp the window to the shorter series via `min(lookback, len(highs), len(lows))`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: backend/models.py [hunk 2]
```diff
@@ -332,6 +332,42 @@ def calculate_atr(
...
+        ratios = [0.0, 0.236, 0.382, 0.5, 0.618, 0.786, 1.0]
+        levels = {str(r): round(swing_high - rng * r, 2) for r in ratios}
+        return {
+            "swing_high": round(swing_high, 2),
+            "swing_low": round(swing_low, 2),
+            "direction": direction,
+            "levels": levels,
+        }
+    except Exception:
+        return None
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply backend/models.py]: Applied — the handler now logs the swallowed exception at DEBUG (`logger.debug("[FIB] ...", exc_info=True)`) before returning None, matching `calculate_atr` and the other indicator helpers.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: frontend/components/PriceChartCard.tsx [hunk 3]
```diff
@@ -56,13 +56,15 @@ export default function PriceChartCard({
   const [intervalData,     setIntervalData]     = useState<IntervalHistoryData | null>(null);
   const [historyLoading,   setHistoryLoading]   = useState(false);
   const [showVwap,         setShowVwap]         = useState(false);
+  const [showFib,          setShowFib]          = useState(false);
   // Track the last symbol we rendered for — reset interval when it changes
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply frontend/components/PriceChartCard.tsx]: Applied — `showFib` is now persisted in localStorage under `fiforesight:overlay:fib` (lazy `useState` initialiser + write-on-toggle, mirroring the SignalPanels pattern) and is no longer reset on symbol change so the preference survives reloads.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: frontend/components/PriceChartCard.tsx [hunk 4]
```diff
@@ -56,13 +56,15 @@ export default function PriceChartCard({
...
   const [lastSymbol, setLastSymbol] = useState(prediction.symbol);
   if (lastSymbol !== prediction.symbol) {
     setLastSymbol(prediction.symbol);
     setSelectedInterval('2y');
     setIntervalData(null);
     setShowVwap(false);
+    setShowFib(false);
```
[the remark this answers is not in the document]
Comment (saikrishna-dev) [reply frontend/components/PriceChartCard.tsx]: Persistence is handled (see `fiforesight:overlay:fib` above). On the setState-in-render point: I kept the symbol-change reset in render. Storing a prev value and conditionally calling setState during render is React’s recommended pattern for resetting state on a prop change (“You Might Not Need an Effect” → “Adjusting some state when a prop changes”); moving it into a `useEffect` adds an extra commit/paint with stale state, so this is intentional rather than an anti-pattern. It is also pre-existing code — this PR only appended the `setShowFib` line, now removed.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
